vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,407 @@
+        such as cv2.approxPolyDP(contour, 0.05*cv2.arcLength(contour, True), True)
+                                          ^^^^can be tweaked
+    test_percent_diff : float
+        The max percent difference in area allowed between the original and approximaed contour
+
+    returns
+    -------
+    returns true if the non-overlapping area between the two contours is less than the specified
+    amount, false otherwise
+    """
+    # kinda need to figure out how much non overlap is too much
+    box: bound_box_type = _min_common_bounding_box((contour, approx))
+
+    contour_mask: mask_type = _generate_mask(contour, box)
+    approx_mask: mask_type = _generate_mask(approx, box)
+
+    non_overlap_mask: mask_type = np.logical_xor(contour_mask, approx_mask)
+
+    non_overlap_cnts: tuple[contour_type]
+    # non_overlap_hier: hierarchy_type
+    non_overlap_img: sc_image_type = non_overlap_mask.astype(UInt8)
+    non_overlap_cnts, _ = cv2.findContours(non_overlap_img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+    non_overlap_area_sum = 0
+    for cnt in non_overlap_cnts:
+        non_overlap_area_sum += cv2.contourArea(cnt)
+
+    contour_area = cv2.contourArea(contour)
+    if (
+        len(non_overlap_cnts) == 0
+        or non_overlap_area_sum == 0
+        or 1 - test_percent_diff < contour_area / non_overlap_area_sum < 1 + test_percent_diff
+    ):
+        return True
+    return False
+
+
+def test_circleness(img: sc_image_type) -> bool:
+    """
+    this function will test the cropped area around a given contour to see if it is a circle or if
+    it is not a circle, it uses the cv2.HoughCircles() function to check for a circle
+
+    parameters
+    ----------
+    img : sc_image_type
+        img should be grayscale cropped box around contour and PROCESSED ALREADY
+        as in the image is binary with the contour filled in as white
+        NOTE: sc_image_type denotes a single channel image
+
+    returns
+    -------
+    returns true if a circle of appropriate size is found (to reduce chance of false positives)
+    """
+    padding: int = int(img.shape[0] * 0.05)
+    modded: sc_image_type = cv2.copyMakeBorder(
+        img, padding, padding, padding, padding, cv2.BORDER_CONSTANT, None, 0
+    )
+    modded = cv2.GaussianBlur(img, (5, 5), 5)
+    circles: circles_type | None = cv2.HoughCircles(
+        modded,
+        cv2.HOUGH_GRADIENT,
+        dp=1,
+        minDist=int(max(img.shape[0], img.shape[1])),
+        param1=50,
+        param2=30,
+        minRadius=4,
+        maxRadius=int(min(img.shape[0], img.shape[1]) * 1.1),
+    )
+
+    if circles is None:
+        return False
+
+    for circle in circles[0]:
+        if (
+            int(max(img.shape[0], img.shape[1]) / 1.6)
+            <= circle[2]
+            <= int(min(img.shape[0], img.shape[1]) * 1.1)
+        ):
+            return True
+
+    return False
+
+
+def filter_contour(
+    contours: tuple[contour_type, ...],
```

Comment:
Similar comment to the one above about indefinite-length tuples. In most cases, it should just be a list.